sockeye/score.py

```diff
@@ -0,0 +1,138 @@
+#
+# or in the "license" file accompanying this file. This file is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+"""
+Scoring CLI.
+"""
+import argparse
+import logging
+import time
+import numpy as np
+from contextlib import ExitStack
+from typing import List, DefaultDict
+
+
+from . import arguments
+from . import constants as C
+from . import utils
+from . import data_io
+from . import translate
+from . import scoring
+
+from sockeye.log import setup_main_logger
+from sockeye.utils import check_condition
+from sockeye.output_handler import get_output_handler, OutputHandler
+
+logger = setup_main_logger(__name__, file_logging=False)
+
+
+def score(output_handler: OutputHandler,
+          models: List[scoring.ScoringModel],
+          data_iters: List[data_io.BaseParallelSampleIter],
+          mapids: List[np.ndarray],
+          scorer: scoring.Scorer) -> None:
+    """
+    Scores all batches returned by all iterators, with all models. Calls
+    an output handler to process the results.
+
+    :param output_handler: A handler that will be called with each scoring result.
+    :param models: A list of models that an instance of scoring.Scorer uses
+    to forward batches.
+    :param data_iters: Iterators that return batches of data.
+    :param mapids: Nested dictionaries mapping  positions in buckets to the
+    original ordering in the input.
+    :param scorer: The scorer that will be used to score each line of input.
+    """
+    logger.info("Scoring...")
+
+    tic = time.time()
+    scored_outputs = scorer.score(models=models,
+                                  data_iters=data_iters,
+                                  mapids=mapids)
+
+    for scored_output in scored_outputs:
+        output_handler.handle(s_output=scored_output)
+
+    total_time = time.time() - tic
+
+    logger.info("Total time taken for scoring: %.4f", total_time)
+
+
+def main():
+    # TODO: test with factors
+    params = argparse.ArgumentParser(description='Scoring CLI')
+
+    arguments.add_score_cli_args(params)
+    args = params.parse_args()
+
+    global logger
+    if args.output is not None:
+        logger = setup_main_logger(__name__,
+                                   console=not args.quiet,
+                                   file_logging=True,
+                                   path="%s.%s" % (args.output, C.LOG_NAME))
+    else:
+        logger = setup_main_logger(__name__,
+                                   console=not args.quiet,
+                                   file_logging=False)
+    utils.log_basic_info(args)
+
+    if args.checkpoints is not None:
+        check_condition(len(args.checkpoints) == len(args.models), "must provide checkpoints for each model")
+
+    with ExitStack() as exit_stack:
+        context = translate._setup_context(args, exit_stack)
```

Comment:
minor: might make sense to remove the underscore to make it explicit that the method is used outside of the module.